docs/architecture-overview.rst

```diff
@@ -0,0 +1,297 @@
+streams, customizations, and so on. This means that one person might
+be a user of multiple Zulip realms. The administrators of a realm can
+choose whether to allow anyone to register an account and join, or
+only allow people who have been invited, or restrict registrations to
+members of particular groups (using email domain names or corporate
+single-sign-on login for verification).
+
+
+
+
+Components
+==========
+
+
+Tornado and Django
+------------------
+
+We use both the `Tornado <http://www.tornadoweb.org>`__ and `Django
+<https://www.djangoproject.com/>`__ Python web frameworks.
+
+Django is the main web application server; Tornado runs the
+server-to-client real-time push system. The app servers are configured
+by the Supervisor configuration (which explains how to start the
+server processes; see "Supervisor" below) and the nginx configuration
+(which explains which HTTP requests get sent to which app server).
+
+Tornado is an asynchronous server, so it can scale really well, but
+it's a real pain to write code for it. Also, we attempt to (as much as
+possible) avoid doing cache or database queries inside the Tornado
+code paths, since those blocking requests carry a very high
+performance penalty for a single-threaded, asynchronous server. So
+Tornado gets only the parts that really need to scale because they
+involve maintaining a persistent connection from every single
+user. The parts that are activated relatively rarely (e.g. when people
+are typing) go into Django. In contrast, websocket requests go through
+Tornado; currently we only use those to improve latency when
+**sending** messages.
+
+
+nginx
+-----
+
+nginx is the front-end to all Zulip activity; it serves static assets
+and proxies to Django and Tornado. It handles HTTP requests according
+to the rules laid down in the many config files found in
+``zulip/puppet/zulip/files/nginx/``.
+
+``zulip/puppet/zulip/files/nginx/zulip-include-frontend/app`` is the
+most important of these files. It explains what happens when requests
+come in from outside.
+
+- In production, all requests to URLs beginning with ``/static/`` are
+  served from the corresponding files in ``/home/zulip/prod-static/``,
+  and the production build process (``link``) compiles, minifies,
+  installs the static assets into the ``prod-static/`` tree form. In
+  development, files are served directly from ``/static/`` in the git
+  repository.
+
+- Requests to ``/json/get_events``, ``/api/v1/events``, and
+  ``/sockjs`` are sent to the Tornado server. These are requests to
+  the real-time push system, because the user's web browser sets up a
+  long-lived TCP connection with Tornado to serve as `a channel for
+  push notifications <
+  https://en.wikipedia.org/wiki/Push_technology#Long_Polling>`__. nginx
+  gets the hostname for the Tornado server via
+  ``puppet/zulip/files/nginx/zulip-include-frontend/upstreams``.
+
+- Requests to ``/api`` and to ``/`` (i.e., everything else) are sent
```

Comment:
This reads as a little confusing -- might be better to just say "everything else" without mentioning `/` and `/api`